README.md
# Website Scraper
Website scraper that returns social graph data

## Prerequisites

You MUST have the following things properly installed and setup on your development device.

* [Git](https://git-scm.com/)
* [Docker](https://docker.com)
* [npm](https://npmjs.com)

## Installing for Development

* `git clone <repository-url>` this repository
* `cd` into the directory
* `docker-compose up` to initialize and build web service and Postgres DB
* `ctrl+c` to stop server. Run `npm install` to load dependencies, and run `docker-compose up` to have the volume with updated files be applied

## Running in Development

All you need is Docker!

To start your development service, `cd {this directory}` and run:

```
docker-compose up
```

You should now be able to access the API service at `http://localhost:8080`.

When making functional changes, push final artifacts to the `master` branch.

## Database
This project uses the Docker postgres image to maintain a local database. When you initially build the docker-compose
project, the database will be seeded with data. If you need to re-seed, you can empty the `./dev-services/postgres/data/pgdata` directory
and run `docker-compose up` again.



changelog.md
# Changelog
All changes made after 05/20/2020 will be documented in this file.

## [05/30/2020]
### Added
 - `getDate()` in place of `setDate()`
 - `makeScrape()` to scrape webpages
 - `makeFrontend()` to create a frontend data object from the data returned by `makeScrape()`
 - query to return "latest crawl within {x} days"
 - comments
 - fixed height for image on frontend

### Removed
 - `setDate()` 
 - Excess axios code
 - Code that deletes properties from `DBdata`
 - JS code that calculated dates

## [05/21/2020]
### Added
- Database rows update if raw_url hasn't been scraped in over a month, rather than adding a new row
- Added a updateScrape function in scrapes.js
- Image and Favicon paths are checked, if they are relative, they are made absolute
- Wrapped the paths in `<a>` tags (on the frontend)
- Cleaned up code and added comments 
- last_modified column
- Trigger that updates the last_modified timestamp when any column is updated

## [05/20/2020]
### Added
 - Promise for getScrape
 - Scraping for url if it hasn't been scraped in over a month

### Removed
 - Callbacks for getScrape and addScrape
 - Excess code

## [Before]
Changes before 05/20/2020 are not logged in this changelog, but you can press [before] to see the commit history.

[05/30/2020]: https://github.com/Gbillington1/WebsiteScraper/compare/5083693..b5b7594 
[05/21/2020]: https://github.com/Gbillington1/WebsiteScraper/compare/57bd12d..6872850
[05/20/2020]: https://github.com/Gbillington1/WebsiteScraper/compare/ef92f98..57bd12d
[Before]: https://github.com/Gbillington1/WebsiteScraper/compare/4cd8ca9..ef92f98

index.js
const axios = require('axios');
const cheerio = require('cheerio');
const http = require('http');
const fs = require('fs');

//create node server, and serve it an html file
var server = http.createServer(function(req, res) {
    console.log("request received!");
    fs.readFile("./index.html", function(err, data) {
        if (err) {
            res.writeHead(404);
            res.write("Not found");
            res.end();
        } else {
            res.writeHead(200, {'Content-Type': 'text/html'});
            res.write(data);
            res.end();
        }
    })
})

server.listen(3000);

//does nothing
$("#submit").click(function() {
    console.log($("#url").val())
})

const url = "http://hugopakula.com";

//use axios get html of url 
axios.get(url)
    .then(function(response) {
        getData(response.data)
    })
    .catch(function(error) {
        console.error(error);
    })

//parse html with cheerio
function getData(html) {
    const $ = cheerio.load(html);
    //find all meta tags with the property of ":og", and log them out in an object
    $('meta[property^="og:"]').each(function() {
        console.log($(this).attr("property") + ": " + $(this).attr("content"));
    })
}



src/models/scrapes.js
// returns all data that match scrapeUrl
function getScrape(client, url) {
    return new Promise(function (resolve, reject) {
        // get data where url == url and has been scraped in the last 28 days
        client.query('SELECT * FROM crawls where raw_url = $1 AND created_at > current_date - interval \'28\' day;', [url], function (err, response) {
            if (err) {
                reject(err)
            } else {
                resolve(response.rows)
            }
        })
    })
}

// adds data to DB
function addScrape(client, url, data) {

    client.query('INSERT INTO crawls (raw_url, title, description, image, favicon) VALUES ($1, $2, $3, $4, $5)', [url, data.title, data.description, data.image, data.favicon])

}

// updates row (rather than making a new row) in DB
function updateScrape(client, url, data) {
    client.query('UPDATE crawls SET title = $1, description = $2, image = $3, favicon = $4 WHERE raw_url = $5', [data.title, data.description, data.image, data.favicon, url])
}

module.exports = {
    'getScrape': getScrape,
    'addScrape': addScrape,
    'updateScrape': updateScrape
};

src/app.js
const express = require('express');
const app = express();
const port = parseInt(process.env.EXPOSED_PORT);
const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs');
const { Client } = require('pg');
const scrapes = require('./models/scrapes');

app.use(express.urlencoded({ extended: true }));

//connect to DB 
const client = new Client({
    connectionString: "postgres://root:pass@postgres:5432/scraper"
});

client
    .connect()
    .then(() => console.log('connected'))
    .catch(err => console.error('error', err.stack));

//serve html page
app.get('/', function (req, res) {
    fs.readFile("./index.html", function (err, data) {
        if (err) {
            res.writeHead(404);
            res.write("Not found");
            res.end();
        } else {
            res.writeHead(200, { 'Content-Type': 'text/html' });
            res.write(data);
            res.end();
        }
    });
});

// make the frontend data with data that is passed in
function makeFrontend(data) {
    let frontendData = {};

    if (data.title) {
        frontendData.title = data.title;
    }

    if (data.description) {
        frontendData.description = data.description;
    }

    if (data.image) {
        frontendData.image = data.image;
    }

    if (data.favicon) {
        frontendData.favicon = data.favicon;
    }

    return frontendData;
}

function makeScrape(scrapeUrl, res) {
    var pattern = /^((http|https|ftp):\/\/)/;
    var DBdata = {};
    //use axios get html of url
    axios.get(scrapeUrl, DBdata)
        .then(function (response) {

            // Load our data into cheerio for DOM parsing
            const $ = cheerio.load(response.data);

            //find all meta tags with the property of ":og", and log them out in an object
            $('meta[property^="og:"]').each(function () {
                switch ($(this).attr("property")) {
                    case 'og:title':

                        DBdata.title = $(this).attr("content");

                        break;
                    case 'og:description':

                        DBdata.description = $(this).attr("content");

                        break;
                    case 'og:image':

                        DBdata.image = $(this).attr('content');

                        // add '/' to the beginning of the link if there isn't one && if image is a relative link
                        if (!(DBdata.image.charAt(0) == '/') && !(pattern.test(DBdata.image))) {
                            DBdata.image = '/' + DBdata.image;
                        }

                        // make the path absolute if it is relative
                        if (!pattern.test(DBdata.image)) {
                            DBdata.image = scrapeUrl + DBdata.image;
                        }
                        break;
                }
            });

            if (typeof DBdata.title === 'undefined') {
                $('head title').each(function () {

                    //if title hasn't been assigned yet
                    if (typeof DBdata.title === 'undefined') {
                        // There should only be one <title> tag, use the first as our title value
                        DBdata.title = $(this).text();
                    }

                });
            }

            //populate the description
            if (typeof DBdata.description === 'undefined') {
                $('p').each(function () {

                    if (typeof DBdata.description === 'undefined') {
                        DBdata.description = $(this).text();
                    }
                })

                if (typeof DBdata.description === 'undefined') {
                    $('h2').each(function () {

                        if (typeof DBdata.description === 'undefined') {
                            DBdata.description = $(this).text();
                        }

                    })
                }

            }

            // populate favicon
            if (typeof DBdata.favicon === 'undefined') {
                //changing from 'head link' to 'head link[rel]' fixed the issue I texted you about, any idea why?
                $('head link[rel]').each(function () {

                    // get link tag that has rel='icon'
                    if ($(this).attr('rel') === 'icon' &&
                        typeof DBdata.favicon === 'undefined') {

                        DBdata.favicon = $(this).attr('href');

                        // add '/' to the beginning of the path if there isn't one
                        if (!(DBdata.favicon.charAt(0) == '/') && !(pattern.test(DBdata.favicon))) {
                            DBdata.favicon = '/' + DBdata.favicon;
                        }

                        // make the path absolute if it is relative
                        if (!pattern.test(DBdata.favicon)) {
                            DBdata.favicon = scrapeUrl + DBdata.favicon;
                        }

                    }

                })
            };

            // add the data to the DB
            scrapes.addScrape(client, scrapeUrl, DBdata)

            return DBdata;
        })
        // take valuable data out of DBdata and make a frontendData object
        .then((DBdata) => {

            var frontendData = makeFrontend(DBdata);

            res.send(frontendData);
        })
        .catch(function (error) {
            console.log(error);
        })
}

//recieve post request
app.post('/scrape', function (req, res) {
    //save URL
    var pattern = /^((http|https|ftp):\/\/)/;
    var scrapeUrl = req.body.url;

    //if url is relative => add 'http://'
    if (!pattern.test(scrapeUrl)) {
        scrapeUrl = "http://" + scrapeUrl;
    }

    // if url ends with '/', remove it 
    if (scrapeUrl.charAt(scrapeUrl.length - 1) === '/') {
        scrapeUrl = scrapeUrl.substring(0, scrapeUrl.length - 1);
    }

    // pull data of scrape url from DB
    scrapes.getScrape(client, scrapeUrl).then(function (rows, error) {
        if (typeof error !== "undefined") {
            res.send({ 'error': true });

            return;
        }

        // if the url hasn't been scraped before (no rows in DB)
        if (rows.length === 0) {

            makeScrape(scrapeUrl, res);

            return;

            // if the url has been scraped before
        } else {
    
            // delete data that the user doesn't care about
            let dataForFrontend = makeFrontend(rows[0]);

            // send data to frontend
            res.send(dataForFrontend);

        }
    }).catch(function (err) {
        console.error(err);
    });


});

// listen for connection on localhost
app.listen(port, () => console.log(`Example app listening at http://localhost:${port}`))